### PR TITLE
Feature/octopoes path queries

### PR DESCRIPTION
**Pull request by Donnype**
### Changes
Adds functionality with which we can build the query we need for https://github.com/minvws/nl-kat-coordination/issues/288 and in this PR: https://github.com/minvws/nl-kat-coordination/pull/761

Integration with the other PR can be done after adding this feature, or we could merge them.

### Issue link
fixes https://github.com/minvws/nl-kat-coordination/issues/836

### Proof
See the tests. I also tested the generated queries on my environment for the use-case and they worked.

---

### Code Checklist
- [x] All the commits in this PR are properly PGP-signed and verified;
- [x] This PR only contains functionality relevant to the issue; tickets have been created for newly discovered issues.
- [x] I have written unit tests for the changes or fixes I made.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

### Communication
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have made corresponding changes to the documentation, if necessary.

---
## Checklist for code reviewers:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---
## Checklist for QA:
Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: octopoes/tests/test_query.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    [ Finding :Finding/ooi Network ]
+    [ Finding :object_type "Finding" ]
+    [ IPPort :IPPort/address IPAddress ]
+    [ IPPort :IPPort/primary_key "IPPort|internet|xxx:xxx:x|tcp|23" ]
+    [ IPPort :object_type "IPPort" ]]}}
+"""
+    )
+
+
+def test_create_query_from_relation_path():
```
Comment (dekkers) [inline octopoes/tests/test_query.py:140]: Test function should also be renamed to `from_path`.

**Review by Lisser — APPROVED** [review]
Neat!

It's difficult to consider all the intricacies that come along with this kind of 'meta'-programming. So I can't be 100% certain that this abstraction 'hits the nail on its head', or that we might find limitations in this abstraction later on. 

One example that comes to mind: we are using the type name (e.g. `Hostname`) as the 'variable-name' in the translated Datalog. So, I wonder what happens if I'd like to make Path query that traverses the same type multiple types. E.g. `Hostname -> DNSTXTRecord -> Hostname -> DNSARecord -> IPAddress`

Does the double Hostname appear as the same symbol in the translated Datalog?

But overall, if this can replace most/all of the plain Datalog queries, then I'd consider this a win already.

File: octopoes/octopoes/xtdb/query.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            if segment.direction is Direction.OUTGOING:
+                query = query.where(segment.source_type, **{segment.property_name: segment.target_type})
+                continue
+
+            query = query.where(segment.target_type, **{segment.property_name: segment.source_type})
+
+        return query
+
+    def query(self, ooi_type: Type[OOI]) -> "Query":
+        """Change the target object type of the Query after initialization, e.g. when using from_relation_path()"""
```
Comment (dekkers) [inline octopoes/octopoes/xtdb/query.py:74]: This doesn't seem to be used anymore, at least not in `from_path`, and also doesn't have test coverage. Is this still needed?

**Comment by Lisser** [comment]
I will try to come up with a test for the edge-case I could think of named above

File: octopoes/octopoes/xtdb/query.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    def from_path(cls, path: Path) -> "Query":
+        """
+        Create a query from a Path.
+
+        The last segment in the path is assumed to be the queries OOI Type. You can change this by calling the
+        query() method after initialization for the required target OOI Type.
+        """
+
+        ooi_type = path.segments[-1].target_type
+        query = Query(ooi_type)
```
Comment (dekkers) [inline octopoes/octopoes/xtdb/query.py:62]: This should be `cls(ooi_type)`.

File: octopoes/octopoes/xtdb/query.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    _offset: Optional[int] = None
+
+    def where(self, ooi_type: Type[OOI], **kwargs) -> "Query":
+        for field_name, value in kwargs.items():
+            self._where_field_is(ooi_type, field_name, value)
+
+        return self
+
+    def format(self) -> str:
+        return "\n" + self._compile(separator="\n    ") + "\n"
```
Comment (dekkers) [inline octopoes/octopoes/xtdb/query.py:50]: Why start the string with a newline?

File: octopoes/octopoes/xtdb/query.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        query() method after initialization for the required target OOI Type.
+        """
+
+        ooi_type = path.segments[-1].target_type
+        query = Query(ooi_type)
+
+        for segment in path.segments:
+            if segment.direction is Direction.OUTGOING:
+                query = query.where(segment.source_type, **{segment.property_name: segment.target_type})
+                continue
```
Comment (dekkers) [inline octopoes/octopoes/xtdb/query.py:67]: I think an `else:` is better here than a `continue` here, because `continue` signals that we skip the rest of the loop for some reason, but that is not really the case.

**Comment by Donnype** [comment]
@Lisser This definitely doesn't cover all of it, but this scenario in particular is a shortcoming indeed: we cannot traverse the same type in multiple places without the query claiming it has to be the _same_ entity. But as this solves the current use-case I think this is something we can iterate on and think of how we want to use the Query object to convey that specific behaviour.

**Comment by Lisser** [comment]
> @Lisser This definitely doesn't cover all of it, but this scenario in particular is a shortcoming indeed: we cannot traverse the same type in multiple places without the query claiming it has to be the same entity. But as this solves the current use-case I think this is something we can iterate on and think of how we want to use the Query object to convey that specific behaviour.

Exactly, good to be aware of it, but doesn't need to fixed before merging this. Maybe good to do create an issue for this though, so we don't forget.

File: octopoes/octopoes/xtdb/query.py (hunk 4 above)
Comment (Donnype) [reply octopoes/octopoes/xtdb/query.py:50]: Was playing around with some formatting in the tests, this looked pretty easy to read. But the newlines at the start and end perhaps don't add mucht?

File: octopoes/octopoes/xtdb/query.py (hunk 2 above)
Comment (Donnype) [reply octopoes/octopoes/xtdb/query.py:74]: Ah perhaps not for now indeed

**Comment by Donnype** [comment]
@Lisser https://github.com/minvws/nl-kat-coordination/issues/911

**Review by Lisser — APPROVED** [review]

**Review by dekkers — APPROVED** [review]
